### ISSUE TITLE
NullReferenceException from SocketAsyncEventArgs.InnerStartOperationReceiveMessageFrom

### ISSUE THREAD

**Issue by lextm** [dotnet/corefx#21995]
I was working on a UDP library when I hit this miserable exception. Attached is the minimal project I can think of to reproduce the exception.

[SocketAsyncClose.zip](https://github.com/dotnet/corefx/files/1132479/SocketAsyncClose.zip)

Compile and run this project with no debugger attached, and you should see the exception details,
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Sockets.SocketAsyncEventArgs.InnerStartOperationReceiveMessageFrom()
   at System.Net.Sockets.Socket.ReceiveMessageFromAsync(SocketAsyncEventArgs e)
   at SocketAsyncClose.Program.<>c__DisplayClass1_0.<Main>b__0() in c:\users\lextm\documents\visual studio 2017\Projects\SocketAsyncClose\SocketAsyncClose\Program.cs:line 36
```
This piece of code tries to reuse `SocketAsyncEventArgs` objects. And it works fine if `Socket.ReceiveAsync` or `Socket.ReceiveFromAsync` is used.

I tested it on Windows 10 (latest patches, .NET Core 1.0 and 1.1), and both cases give the same exception. I tested on macOS (latest stable. .NET Core 1.0 and 1.1), and cannot reproduce the exception. 

Not surprised, as it should be something wrong in `SocketAsyncEventArgs.Windows.cs` I think and is Windows only.

**Comment by lextm** [comment]
The severity of this issue is high, due to
- `Socket.ReceiveMessageFromAsync` is the only async way to get the remote endpoint address, per the documentation. So, other async receive methods cannot be used instead.
- The `SocketAsyncEventArgs` served becomes wild after this exception, and cannot be reused (if reused, an `InvalidOperationException` would occur, which is used by this sample project to terminate execution).  Recreating an instance of `SocketAsyncEventArgs` would be expensive, and there is no easy way to track which instance is wild and must be recreated.

**Comment by davidsh** [comment]
Does this repro on .NET Framework?

**Comment by lextm** [comment]
@davidsh Nope. It does not reproduce if I use the same code in a .NET Framework 4.7 console application on Windows 10.

**Comment by stephentoub** [comment]
I believe this was already fixed (I think in https://github.com/dotnet/corefx/pull/17237).  Can you try .NET Core 2.0?  With your repro, I get the exception on 1.1 but not on 2.0.

**Comment by lextm** [comment]
Thanks @stephentoub . 

I revised the sample as attached.
[SocketAsyncClose2.zip](https://github.com/dotnet/corefx/files/1133022/SocketAsyncClose2.zip)

**.NET Framework 4.7 on Windows 10**
The output can be used as baseline,
```
Press any key to close the socket.
127.0.0.1:54113
received 0
Sent 0
127.0.0.1:54114
received 1
Sent 1
127.0.0.1:54115
received 2
Sent 2
<-- empty line as I hit Enter key
0.0.0.0:555
received 0
Press any key to close the program.
```
**.NET Core on macOS**
It gives exactly the same result like .NET Framework on Windows 10.

The .NET Core builds I tested are,

- 1.0.5
- 1.1.2
- 2.0 preview 2

**.NET Core on Windows 10**
.NET Core 1.x shows the exception,
```
Press any key to close the socket.
Sent 0
127.0.0.1:49946
received 0
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Sockets.SocketAsyncEventArgs.InnerStartOperationReceiveMessageFrom()
   at System.Net.Sockets.Socket.ReceiveMessageFromAsync(SocketAsyncEventArgs e)
   at SocketAsyncClose.Program.<>c__DisplayClass1_0.<Main>b__0() in C:\Users\lextm\documents\visual studio 2017\Projects\SocketAsyncClose\SocketAsyncClose\Program.cs:line 36
break from invalid operation exception
Press any key to continue . . .
```

Then I tested several .NET Core 2.x runtime versions,.

- 2.0.0-preview2-25407-01
- 2.0.0-preview3-25507-02
- 2.1.0-preview2-25508-03

Their output is strange,
```
Press any key to close the socket.
Sent 0
Sent 1
Sent 2

Press any key to close the program.
```
**Yes. There is no more exception, but the packets are not received at all. Any idea why?** 

If it is a new bug, then I am OK we close this one and open a new one to track.

**Comment by stephentoub** [comment]
Thanks, @lextm.

It looks like the bug is actually still there.  I see the problem and am looking at a fix.

But your repro also has a bug in it.  ReceiveMessageFromAsync may complete synchronously, but your code isn't handling that case, which is why you're seeing a hang.

**Comment by lextm** [comment]
Thank you @stephentoub for the tip.

[SocketAsyncClose3.zip](https://github.com/dotnet/corefx/files/1133522/SocketAsyncClose3.zip)

I revised the sample project again to include sync receive handling, and then the output for different scenarios become clearer.

For scenarios not affected by this bug, the output is generally the same,
* .NET Framework 4.7 on Windows 10
* .NET Core on macOS
```
Press any key to close the socket.
127.0.0.1:59601
async received 0
Sent 0
127.0.0.1:59602
async received 1
Sent 1
127.0.0.1:59603
async received 2
Sent 2

0.0.0.0:555
async received 0
Press any key to close the program.
```
The exception can be reproduced with .NET Core 1.x on Windows, and the output is
```
Press any key to close the socket.
127.0.0.1:60826
Sent 0
async received 0 <-- it is async here
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Sockets.SocketAsyncEventArgs.InnerStartOperationReceiveMessageFrom()
   at System.Net.Sockets.Socket.ReceiveMessageFromAsync(SocketAsyncEventArgs e)
   at SocketAsyncClose.Program.<>c__DisplayClass1_0.<Main>b__0() in C:\Users\lextm\documents\visual studio 2017\Projects\SocketAsyncClose\SocketAsyncClose\Program.cs:line 36
break from invalid operation exception
```
The exception can be also reproduced with .NET Core 2.x on Windows, the output changes to,
```
Press any key to close the socket.
Sent 0
127.0.0.1:57465
sync received 0 <-- It is sync here.
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Sockets.SocketAsyncEventArgs.InnerStartOperationReceiveMessageFrom()
   at System.Net.Sockets.Socket.ReceiveMessageFromAsync(SocketAsyncEventArgs e)
   at SocketAsyncClose.Program.<>c__DisplayClass1_0.<Main>b__0() in C:\Users\lextm\Downloads\test\Program.cs:line 36
break from invalid operation exception
```

**Comment by stephentoub** [comment]
Fixed in #22012 and #22013.

**Comment by lextm** [comment]
Thanks. Glad to see it would land on .NET Core 2 RTM.
